### ISSUE TITLE
Data Problems and Questions

### ISSUE THREAD

**Issue by tonofshell** [liu431/Big-Data-Project#10]
I have found a few issues that will add to the complexity of our analysis that I think we should start thinking about. Feel free to add any questions or problems you find with the data to this issue.

- The data is in XML format *__NOT__* raw text files
    - Can we parse XML line by line?
    - Depending on the XML parser, this is likely stored in Python as a list of dictionaries, one dictionary for each row in the XML file. 
        - Each row is one user or comment
- Any comment or About Me attribute (basically anything with a sentence or more of text) within each XML row is in HTML format
    - Do we simply drop all HTML tags and how would we do this?
    - Do we process the data and save it to disk in another format to use for our analysis?

**Comment by sanittawan** [comment]
Thanks for laying out all the questions. To answer some of your questions:

* The data is in XML format NOT raw text files
    - Can we parse XML line by line? >> I think we can use `lxml` or `beautifulsoup` to do it. Check [this link](https://www.ibm.com/developerworks/xml/library/x-hiperfparse/) and this [StackOverflow thread](https://codereview.stackexchange.com/questions/194703/importing-xml-dump-of-stack-overflow-questions-and-answers-into-sqlite3) out. 
    - Depending on the XML parser, this is likely stored in Python as a list of dictionaries, one dictionary for each row in the XML file. >> I think that we can parse and store it in SQLite. I am awaiting for Prof. Wachs' answer on this. 

* Any comment or About Me attribute (basically anything with a sentence or more of text) within each XML row is in HTML format
    - Do we simply drop all HTML tags and how would we do this? >> We can use `beautifulsoup` to extract the text, theoretically.
    - Do we process the data and save it to disk in another format to use for our analysis? >> Any ideas?

**Comment by tonofshell** [comment]
Just an answer to another question that I hadn't proposed yet, how much will it cost to store the data in Google Cloud? To store it on a drive for access by a compute cluster, we are subject to [this pricing scheme](https://cloud.google.com/compute/pricing?hl=en_US#persistentdisk):
![2019-04-29 (3)](https://user-images.githubusercontent.com/1500631/56930552-a2cb7600-6aa2-11e9-88d2-2d93aba68aac.png)

We can store it in a Google Cloud storage bucket which is [cheaper for just storing the data](https://cloud.google.com/storage/pricing), and it [appears we can access it from compute](https://cloud.google.com/compute/docs/disks/gcs-buckets), but then we would need to pay a cost per operation (query) as well.

**Comment by sanittawan** [comment]
Regarding @tonofshell 's comment on storage, I chatted with Prof. Wachs today and he said that we should store the data in Google cloud buckets (ref. [here](https://cloud.google.com/storage/docs/creating-buckets) and [here](https://cloud.google.com/storage/docs/key-terms#buckets). Students in the past years store their data there and access it using [gsutil tool](https://cloud.google.com/storage/docs/gsutil).

Also to note, we can automate the launching of compute engines via [gcloud command](https://cloud.google.com/pubsub/docs/quickstart-cli).

**Comment by tonofshell** [comment]
> Regarding @tonofshell 's comment on storage, I chatted with Prof. Wachs today and he said that we should store the data in Google cloud buckets (ref. [here](https://cloud.google.com/storage/docs/creating-buckets) and [here](https://cloud.google.com/storage/docs/key-terms#buckets). Students in the past years store their data there and access it using [gsutil tool](https://cloud.google.com/storage/docs/gsutil).
> 
> Also to note, we can automate the launching of compute engines via [gcloud command](https://cloud.google.com/pubsub/docs/quickstart-cli).

I guess this means we need to decide __how much__ of the data we want to use before we start paying to put all 200+Gb into a Google Cloud Bucket

**Comment by sanittawan** [comment]
> > Regarding @tonofshell 's comment on storage, I chatted with Prof. Wachs today and he said that we should store the data in Google cloud buckets (ref. [here](https://cloud.google.com/storage/docs/creating-buckets) and [here](https://cloud.google.com/storage/docs/key-terms#buckets). Students in the past years store their data there and access it using [gsutil tool](https://cloud.google.com/storage/docs/gsutil).
> > Also to note, we can automate the launching of compute engines via [gcloud command](https://cloud.google.com/pubsub/docs/quickstart-cli).
> 
> I guess this means we need to decide **how much** of the data we want to use before we start paying to put all 200+Gb into a Google Cloud Bucket

Exactly. Prof Wachs gave a piece of advice that we should start from <1% of the entire data, store it one the cloud buckets, and do some analysis. Once we are sure that our code won't break, we can either slowly increase the size of the data (from <1% to 5% to 10% to 20% and so on) or we can try putting the entire 200 GB there. He did mention though that the storage won't burn our money quickly. It's the computation that will cost us a lot. He also reminded us to select a compute engine with 500 GB when we want to upload full data set to cloud.
